### PR TITLE
refactor: Cleaned pyroengine and refactored project

### PR DESCRIPTION
**Pull request by frgfm**
This PR introduces the following changes:
- simplified `pyroengine`
    - deprecates `pi_utils` (not used)
    - moved `pyroengine.engine.engine` to `pyroengine.core`
    - added `pyroengine.vision` with `Classifier` which holds the ONNX model
- updated `tests`
    - moved "test" to "tests"
    - switched from unittests to pytest to harmonize our test suite across repos
    - added exhaustive unit tests for pyroengine (only part not being tested : API interactions, because it requires to run it locally)
- updated `docs`
    - updated documentation theme from rtd_theme to furo (cf. modifications in pyrovision)
- deleted `server` folder
- renamed `runner` as `src`
- cleaned `src` by implementing `ReolinkCamera` and `SystemController`
- refactored all requirements.txt and config files into `pyproject.toml`
- added a `Makefile`
- updated CI jobs accordingly
- updated docker file accordingly

Basically, for the wrapper (orginially `runner`, now `src`), the only difference is:

- previously, we needed the following keys: "model_weights"
- now, we need: "cfg_path" (describing the preprocessing), "model_path" (path to ONNX model), "hub_repo" (fallback to download the model from HF hub)

For reviewers, I highly suggest to check the modifications by commit, it might be easier :sweat_smile:

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    "Programming Language :: Python :: 3.7",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Topic :: Scientific/Engineering",
+    "Topic :: Scientific/Engineering :: Mathematics",
+    "Topic :: Scientific/Engineering :: Artificial Intelligence",
+]
+dynamic = ["version"]
+dependencies = [
+    "Pillow>=8.4.0",
```
Comment (MateoLostanlen) [inline pyproject.toml:33]: it's not better to use exact versions for deployment ?

File: pyroengine/core.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+            self._dump_cache()
+            self.last_cache_dump = ts
+
+        # save frame
+        if len(self.api_client) > 0 and isinstance(self.frame_saving_period, int) and isinstance(cam_id, str):
+            self.frames_counter[cam_id] += 1
+            if self.frames_counter[cam_id] == self.frame_saving_period:
+                # Send frame to the api
+                stream = io.BytesIO()
+                frame.save(stream, format="JPEG")
```
Comment (MateoLostanlen) [inline pyroengine/core.py:228]: maybe it's not the time to do it but jpeg default compression is 75% for jpeg, it's a bit low

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+on:
+  push:
+    branches: main
+  pull_request:
+    branches: main
+
+jobs:
+  package:
+    runs-on: ubuntu-latest
```
Comment (MateoLostanlen) [inline .github/workflows/docker.yml:11]: The goal here is to build an image to test on your pc ? You are crushing the image used by the towers that must be built for arm. We have to setup a cross-build but we can't overwrite the existing image

File: pyroengine/__init__.py [hunk 4]
```diff
@@ -1,6 +1,2 @@
-# Copyright (C) 2020-2022, Pyronear.
```
Comment (MateoLostanlen) [inline pyroengine/__init__.py:1]: Why do you remove the header here ?

**Review by MateoLostanlen — CHANGES_REQUESTED** [review]
Hi @fg, thanks so much for this PR! Globally LGTM. Several sticking points though:

1. This PR must be merge on develop and not main directly
2. CIs must run on main & develop

To discuss

We never used the felix code to monitor the pi's but I think it would be nice to do so, I don't know if it's a good idea to delete it

NB: It would be better to use a fork

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+keywords = ["pytorch", "deep learning", "vision", "cnn", "wildfire"]
+classifiers = [
+    "Development Status :: 4 - Beta",
+    "Intended Audience :: Developers",
+    "Intended Audience :: Science/Research",
+    "License :: OSI Approved :: Apache Software License",
+    "Natural Language :: English",
+    "Operating System :: OS Independent",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.6",
```
[the remark this answers is not in the document]
Comment (frgfm) [reply pyproject.toml:23]: Well the only thing constraining the version of Python in Python3 is the use of f-string
Subdependencies might have subtilities, but our codebase is fully compatible with 3.6 & higher :)

File: pyproject.toml (hunk 1 above)
Comment (frgfm) [reply pyproject.toml:33]: Yes it is, but I figured the package should have loose specifiers, and `src/requirements.txt` should have fixed ones :)

File: pyroengine/core.py (hunk 2 above)
Comment (frgfm) [reply pyroengine/core.py:228]: Mmmh the option to downsize the image is already in place, we can check this in another PR (since we'll need some benchmarking)

File: pyroengine/__init__.py (hunk 4 above)
Comment (frgfm) [reply pyroengine/__init__.py:1]: files with no added value code (`__init__.py` and `version.py`) have no reason to include the header 
It's just adding commented sections in files that have barely anything

File: .github/workflows/docker.yml (hunk 3 above)
Comment (frgfm) [reply .github/workflows/docker.yml:11]: ? I'm not sure what you mean: I haven't removed anything here, only added a test.

To me, we have to make sure that the image can at least be built on ubuntu. If it doesn't, we can expect it to fail on ARM. Now, I wanted to add also a CI to build the image on ARM but I have no clue how long it takes on RPI. I suggest to add this in another PR along with the orchestration of the CI to include API interactions :+1:

File: pyproject.toml (hunk 1 above)
Comment (frgfm) [reply pyproject.toml:33]: Actually, let me set poetry in place for the `src`, that will help us

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        with:
+          python-version: 3.8
+          architecture: x64
+      - name: Install package
+        run: |
+          python -m pip install --upgrade pip
+          pip install pyroengine
+          python -c "import pyroengine; print(pyroengine.__version__)"
+
+  dockerhub-publish:
```
Comment (MateoLostanlen) [inline .github/workflows/release.yml:58]: My bad the push is here

File: .github/workflows/docker.yml (hunk 3 above)
Comment (MateoLostanlen) [reply .github/workflows/docker.yml:11]: You erase the [current image](https://hub.docker.com/repository/registry-1.docker.io/pyronear/pyro-engine/tags?page=1&ordering=last_updated) on hub but pushing a linux/arm64 image, but the docker pushis not here sorry !

Comment (MateoLostanlen) [reply .github/workflows/docker.yml:11]: https://github.com/pyronear/pyro-engine/blob/ea017e1cff156f6d1137cd7684180288b7aa8a21/.github/workflows/release.yml#L58

Comment (MateoLostanlen) [reply .github/workflows/docker.yml:11]: But yes we can fix it manually and managed this in a comming PR

File: pyproject.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (MateoLostanlen) [reply pyproject.toml:23]: ok ok

File: pyroengine/core.py (hunk 2 above)
Comment (MateoLostanlen) [reply pyroengine/core.py:228]: Yes I just wanted to raise the "issue", we can manage this later with a proper bench

File: pyproject.toml (hunk 1 above)
Comment (frgfm) [reply pyproject.toml:33]: will be resolved in #102

**Comment by MateoLostanlen** [comment]
> Hi @fg, thanks so much for this PR! Globally LGTM. Several sticking points though:
> 
>     1. This PR must be merge on develop and not main directly
> 
>     2. CIs must run on main & develop
> 
> 
> To discuss
> 
> We never used the felix code to monitor the pi's but I think it would be nice to do so, I don't know if it's a good idea to delete it
> 
> NB: It would be better to use a fork

Thanks for the update, what about these comments ?

**Comment by frgfm** [comment]
> Thanks for the update, what about these comments ?

Oops I didnt see the last part:
- I removed it because it comes with dependencies (I would have kept it otherwise)
- thought it would be better to start with a clean environment, then add it back later and see how we integrate it

**Review by MateoLostanlen — APPROVED** [review]
LGTM
